rasterio/env.py

```diff
@@ -148,19 +148,24 @@ def __init__(
             raise EnvError(
                 "GDAL's AWS config options can not be directly set. "
                 "AWS credentials are handled exclusively by boto3.")
-        self.aws_unsigned = aws_unsigned
-        self.aws_access_key_id = aws_access_key_id
-        self.aws_secret_access_key = aws_secret_access_key
-        self.aws_session_token = aws_session_token
-        self.region_name = region_name
-        self.profile_name = profile_name
-        self.session = session
+
+        if session:
+            # Special case for older Rasterio usage.
+            if not isinstance(session, Session):
+                session = AWSSession(session=session)
+            self.session = session
+        else:
```

Comment:
TODO: note that this is also to support older usage.

TODO: send `RasterioDeprecationWarning` once this gets merged to master.